File: nora-backend/app/noraops/services/pyproject_resolve.py
"""pyproject.toml 探索・優先度（サーバー側リポ走査 / 監査用）。

探索範囲と優先度は vscode-extension/src/noraops/pyprojectResolve.js と同期すること。
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

PYPROJECT_FILE = "pyproject.toml"
LEGACY_PACKAGES_REL = Path("nora/packages")

SKIP_SUBDIRS = frozenset(
    {
        ".git",
        ".nora",
        ".venv",
        "venv",
        "node_modules",
        "__pycache__",
        "dist",
        "build",
        ".vscode",
        "static",
        "media",
        "logs",
        "assets",
    }
)

PREFERRED_SUBDIRS = ("app", "backend", "server", "src", "api", "web", "python")

PRIORITY_MANIFEST = 0
PRIORITY_ROOT = 10
PRIORITY_LEGACY = 20
PRIORITY_SHALLOW_BASE = 30


@dataclass(frozen=True)
class PyprojectCandidate:
    dir_rel: str
    pyproject_rel: str
    source: str
    priority: int


@dataclass(frozen=True)
class PyprojectResolution:
    ok: bool
    app_root: Path
    project_dir: Path
    pyproject_path: Path
    pyproject_rel: str
    source: str | None
    candidates: tuple[PyprojectCandidate, ...]
    ambiguous: bool
    alternate_rels: tuple[str, ...]


def _normalize_rel_posix(rel: str | None) -> str:
    if not rel or rel in (".", "./"):
        return "."
    return str(rel).replace("\\", "/").rstrip("/")


def packages_project_rel(manifest: dict | None) -> str:
    if not manifest:
        return "."
    raw = manifest.get("packagesProject")
    if raw is None or raw == "" or raw == ".":
        return "."
    return _normalize_rel_posix(str(raw))


def _project_dir_from_rel(app_root: Path, dir_rel: str) -> Path:
    if dir_rel == ".":
        return app_root
    return app_root / dir_rel


def _pyproject_rel_from_dir_rel(dir_rel: str) -> str:
    if dir_rel == ".":
        return PYPROJECT_FILE
    return f"{dir_rel}/{PYPROJECT_FILE}"


def _shallow_subdir_priority(name: str) -> int:
    lower = name.lower()
    try:
        return PREFERRED_SUBDIRS.index(lower)
    except ValueError:
        return 1000


def _read_manifest(app_root: Path) -> dict | None:
    man_path = app_root / "nora" / "manifest.json"
    if not man_path.is_file():
        return None
    try:
        return json.loads(man_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None


def discover_pyproject_candidates(
    workspace_root: str | Path,
    *,
    manifest: dict | None | object = _read_manifest,
) -> list[PyprojectCandidate]:
    app_root = Path(workspace_root).resolve()
    if not app_root.is_dir():
        return []

    if manifest is _read_manifest:
        manifest = _read_manifest(app_root)
    elif manifest is None:
        manifest = None

    candidates: list[PyprojectCandidate] = []
    seen: set[str] = set()

    def push(dir_rel: str, source: str, priority: int) -> None:
        norm = _normalize_rel_posix(dir_rel)
        if norm in seen:
            return
        project_dir = _project_dir_from_rel(app_root, norm)
        py_path = project_dir / PYPROJECT_FILE
        if not py_path.is_file():
            return
        seen.add(norm)
        candidates.append(
            PyprojectCandidate(
                dir_rel=norm,
                pyproject_rel=_pyproject_rel_from_dir_rel(norm),
                source=source,
                priority=priority,
            )
        )

    if manifest and manifest.get("packagesProject") is not None:
        raw = str(manifest.get("packagesProject", "")).strip()
        if raw:
            push(packages_project_rel(manifest), "manifest", PRIORITY_MANIFEST)

    push(".", "root", PRIORITY_ROOT)
    push(str(LEGACY_PACKAGES_REL).replace("\\", "/"), "legacy", PRIORITY_LEGACY)

    try:
        entries = list(app_root.iterdir())
    except OSError:
        entries = []

    for ent in entries:
        if not ent.is_dir():
            continue
        name = ent.name
        if name in SKIP_SUBDIRS or name == "nora":
            continue
        sub_priority = PRIORITY_SHALLOW_BASE + _shallow_subdir_priority(name)
        push(name, "shallow", sub_priority)

    candidates.sort(key=lambda c: (c.priority, c.dir_rel))
    return candidates


def resolve_pyproject(
    workspace_root: str | Path,
    *,
    manifest: dict | None | object = _read_manifest,
) -> PyprojectResolution:
    app_root = Path(workspace_root).resolve()
    candidates = discover_pyproject_candidates(workspace_root, manifest=manifest)

    if not candidates:
        return PyprojectResolution(
            ok=False,
            app_root=app_root,
            project_dir=app_root,
            pyproject_path=app_root / PYPROJECT_FILE,
            pyproject_rel=PYPROJECT_FILE,
            source=None,
            candidates=(),
            ambiguous=False,
            alternate_rels=(),
        )

    chosen = candidates[0]
    project_dir = _project_dir_from_rel(app_root, chosen.dir_rel)
    alts = tuple(c.pyproject_rel for c in candidates[1:])

    return PyprojectResolution(
        ok=True,
        app_root=app_root,
        project_dir=project_dir,
        pyproject_path=project_dir / PYPROJECT_FILE,
        pyproject_rel=chosen.pyproject_rel,
        source=chosen.source,
        candidates=tuple(candidates),
        ambiguous=len(candidates) > 1,
        alternate_rels=alts,
    )


def read_pyproject_text(workspace_root: str | Path, *, manifest: dict | None | object = _read_manifest) -> str | None:
    res = resolve_pyproject(workspace_root, manifest=manifest)
    if not res.ok:
        return None
    try:
        return res.pyproject_path.read_text(encoding="utf-8")
    except OSError:
        return None


def parse_pyproject_summary(pyproject_path: str | Path) -> dict:
    """pyproject.toml から requires-python / dependencies 名を抽出（監査用）。"""
    path = Path(pyproject_path)
    out: dict = {"requiresPython": None, "dependencies": []}
    if not path.is_file():
        return out
    try:
        import tomllib

        data = tomllib.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return out

    project = data.get("project") or {}
    rp = project.get("requires-python") or project.get("requires_python")
    if rp:
        out["requiresPython"] = str(rp)

    deps: list[str] = []
    for item in project.get("dependencies") or []:
        name = str(item).strip().split("[", 1)[0].strip()
        for sep in ("==", ">=", "<=", "~=", "!=", "<", ">"):
            if sep in name:
                name = name.split(sep, 1)[0].strip()
                break
        if name:
            deps.append(name)
    out["dependencies"] = deps
    return out

File: vscode-extension/src/noraops/pyprojectResolve.js
/**
 * pyproject.toml の探索・優先度（Creator / Runner / ポリシー共通）
 *
 * 探索範囲: ワークスペースルート + 直下 1 階層（レガシー nora/packages は互換で明示）
 * 優先度:
 *   1. nora/manifest.json packagesProject（存在する場合）
 *   2. ルート pyproject.toml
 *   3. nora/packages/pyproject.toml（旧レイアウト）
 *   4. 直下サブディレクトリ（app, backend, src … → その他は名前順）
 *
 * @see NoraOps/互換方針.md
 */

const fs = require("fs");
const path = require("path");
const { resolveScaffoldRoot } = require("./scaffold");

const PYPROJECT_FILE = "pyproject.toml";
const LEGACY_PACKAGES_REL = "nora/packages";

/** 探索しない直下フォルダ */
const SKIP_SUBDIRS = new Set([
  ".git",
  ".nora",
  ".venv",
  "venv",
  "node_modules",
  "__pycache__",
  "dist",
  "build",
  ".vscode",
  "static",
  "media",
  "logs",
  "assets",
]);

/** 同階層で複数あるときの優先サブディレクトリ名 */
const PREFERRED_SUBDIRS = ["app", "backend", "server", "src", "api", "web", "python"];

const PRIORITY = {
  manifest: 0,
  root: 10,
  legacy: 20,
  shallowBase: 30,
};

function normalizeRelPosix(rel) {
  if (!rel || rel === "." || rel === "./") return ".";
  return String(rel).replace(/\\/g, "/").replace(/\/+$/, "");
}

function packagesProjectRel(manifest) {
  const raw = manifest?.packagesProject;
  if (raw == null || raw === "" || raw === ".") return ".";
  return normalizeRelPosix(raw);
}

function projectDirFromRel(appRoot, rel) {
  if (rel === ".") return appRoot;
  return path.join(appRoot, rel);
}

function pyprojectRelFromDirRel(dirRel) {
  return dirRel === "." ? PYPROJECT_FILE : `${dirRel}/${PYPROJECT_FILE}`;
}

function shallowSubdirPriority(name) {
  const lower = name.toLowerCase();
  const idx = PREFERRED_SUBDIRS.indexOf(lower);
  return idx >= 0 ? idx : 1000;
}

/**
 * @param {string} workspaceRoot
 * @param {{ manifest?: object | null, readManifest?: boolean }} [options]
 * @returns {Array<{ dirRel: string, pyprojectRel: string, source: string, priority: number }>}
 */
function discoverPyprojectCandidates(workspaceRoot, options = {}) {
  const appRoot = resolveScaffoldRoot(workspaceRoot);
  if (!appRoot || !fs.existsSync(appRoot)) return [];

  let manifest = options.manifest;
  if (manifest === undefined && options.readManifest !== false) {
    try {
      const { readNoraManifest } = require("./appEntry");
      manifest = readNoraManifest(appRoot);
    } catch {
      manifest = null;
    }
  }

  const candidates = [];
  const seen = new Set();

  const push = (dirRel, source, priority) => {
    const norm = normalizeRelPosix(dirRel);
    const dir = projectDirFromRel(appRoot, norm);
    const py = path.join(dir, PYPROJECT_FILE);
    if (!fs.existsSync(py)) return;
    const key = norm;
    if (seen.has(key)) return;
    seen.add(key);
    candidates.push({
      dirRel: norm,
      pyprojectRel: pyprojectRelFromDirRel(norm),
      source,
      priority,
    });
  };

  if (manifest?.packagesProject != null && String(manifest.packagesProject).trim() !== "") {
    push(packagesProjectRel(manifest), "manifest", PRIORITY.manifest);
  }

  push(".", "root", PRIORITY.root);
  push(LEGACY_PACKAGES_REL, "legacy", PRIORITY.legacy);

  let entries;
  try {
    entries = fs.readdirSync(appRoot, { withFileTypes: true });
  } catch {
    entries = [];
  }

  for (const ent of entries) {
    if (!ent.isDirectory()) continue;
    if (SKIP_SUBDIRS.has(ent.name)) continue;
    if (ent.name === "nora") continue;
    const subPriority = PRIORITY.shallowBase + shallowSubdirPriority(ent.name);
    push(ent.name, "shallow", subPriority);
  }

  candidates.sort((a, b) => {
    if (a.priority !== b.priority) return a.priority - b.priority;
    return a.dirRel.localeCompare(b.dirRel);
  });

  return candidates;
}

/**
 * @param {string} workspaceRoot
 * @param {{ manifest?: object | null, readManifest?: boolean }} [options]
 */
function resolvePyproject(workspaceRoot, options = {}) {
  const appRoot = resolveScaffoldRoot(workspaceRoot);
  const candidates = discoverPyprojectCandidates(workspaceRoot, options);

  if (!candidates.length) {
    return {
      ok: false,
      appRoot,
      projectDir: appRoot,
      pyprojectPath: path.join(appRoot, PYPROJECT_FILE),
      pyprojectRel: PYPROJECT_FILE,
      source: null,
      candidates: [],
      ambiguous: false,
    };
  }

  const chosen = candidates[0];
  const projectDir = projectDirFromRel(appRoot, chosen.dirRel);

  return {
    ok: true,
    appRoot,
    projectDir,
    pyprojectPath: path.join(projectDir, PYPROJECT_FILE),
    pyprojectRel: chosen.pyprojectRel,
    source: chosen.source,
    candidates,
    ambiguous: candidates.length > 1,
    alternateRels: candidates.slice(1).map((c) => c.pyprojectRel),
  };
}

function hasResolvedPyproject(workspaceRoot, options = {}) {
  return resolvePyproject(workspaceRoot, options).ok;
}

module.exports = {
  PYPROJECT_FILE,
  LEGACY_PACKAGES_REL,
  SKIP_SUBDIRS,
  PREFERRED_SUBDIRS,
  PRIORITY,
  packagesProjectRel,
  discoverPyprojectCandidates,
  resolvePyproject,
  hasResolvedPyproject,
};
